src/main/java/bio/terra/filesystem/FireStoreDao.java

```diff
@@ -0,0 +1,254 @@
+        String datasetId = dataset.getId().toString();
+        return directoryDao.validateRefIds(firestore, datasetId, refIdArray);
+    }
+
+    // -- private methods --
+
+    private FSObjectBase retrieveWorker(Firestore firestore,
+                                        String datasetId,
+                                        int enumerateDepth,
+                                        FireStoreObject fireStoreObject,
+                                        boolean throwOnNotFound,
+                                        String context) {
+        if (fireStoreObject == null) {
+            return handleNotFound(throwOnNotFound, context);
+        }
+
+        if (fireStoreObject.getFileRef()) {
+            FSObjectBase fsFile = makeFSFile(firestore, datasetId, fireStoreObject);
+            if (fsFile == null) {
+                // We found a file in the directory that is not done being created. We treat this
+                // as not found.
+                return handleNotFound(throwOnNotFound, context);
+            }
+            return fsFile;
+        }
+        return makeFSDir(firestore, datasetId, enumerateDepth, fireStoreObject);
+    }
+
+    private FSObjectBase handleNotFound(boolean throwOnNotFound, String context) {
+        if (throwOnNotFound) {
+            throw new FileSystemObjectNotFoundException("File system object not found: " + context);
+        } else {
+            return null;
+        }
+    }
+
+    private FSObjectBase makeFSDir(Firestore firestore,
+                                   String datasetId,
+                                   int level,
+                                   FireStoreObject fireStoreObject) {
+        if (fireStoreObject.getFileRef()) {
+            throw new IllegalStateException("Expected directory; got file!");
+        }
+
+        String fullPath = fireStoreUtils.getFullPath(fireStoreObject.getPath(), fireStoreObject.getName());
+
+        FSDir fsDir = new FSDir();
+        fsDir
+            .objectId(UUID.fromString(fireStoreObject.getObjectId()))
+            .datasetId(UUID.fromString(datasetId))
+            .createdDate(Instant.parse(fireStoreObject.getFileCreatedDate()))
+            .path(fullPath)
+            .checksumCrc32c(fireStoreObject.getChecksumCrc32c())
+            .checksumMd5(fireStoreObject.getChecksumMd5())
+            .size(fireStoreObject.getSize())
+            .description(StringUtils.EMPTY);
+
+
+        if (level != 0) {
+            List<FSObjectBase> fsContents = new ArrayList<>();
+            List<FireStoreObject> dirContents =
+                directoryDao.enumerateDirectory(firestore, datasetId, fireStoreObject.getPath());
+            for (FireStoreObject fso : dirContents) {
+                if (fso.getFileRef()) {
+                    // Skip files that are not fully created
+                    FSObjectBase fsFile = makeFSFile(firestore, datasetId, fso);
```

Comment:
The new file load flight uses the sequence of writes to know when a file is finished being loaded, instead of updating a status field in the data. The sequence is:
1. Create the directory entry - that transaction prevents a second, duplicately named directory entry from being made.
2. Load the file into GCS, getting the file info (checksum, size, etc)
3. Create the complete file entry
So if we are enumerating after step 1 and before step 3, we would see a directory entry for a file, but the file isn't all there yet.

I'll cover this in the design doc!